src/libflake/flake/flake.cc

```diff
@@ -947,8 +947,10 @@ std::optional<Fingerprint> LockedFlake::getFingerprint(ref<Store> store) const
 {
     if (lockFile.isUnlocked()) return std::nullopt;
 
-    auto fingerprint = flake.lockedRef.input.getFingerprint(store);
-    if (!fingerprint) return std::nullopt;
+    auto fingerprint = flake.lockedRef.input.getFingerprint(store)
+        // Fingerprints are null for relative path references so we use
+        // the flake path as a fallback to avoid losing caching.
+        .value_or(flake.path.to_string());
```

Comment:
This is either safe by accident or unsafe, because this undocumented :anger:  `to_string()` is calling methods that are for display and not guaranteed to be a functional key for the sources it represents.
However, I can imagine `flake.path` here to be a store path, in which case we could use it after making sure that it is one.

Reply:
> That's intentional because computing a fingerprint for path flakes requires reading the entire flake.

That seems to apply here?

Reply:
FYI this `to_string()` is exactly how fingerprints used to be calculated 😅 

I will remove this part of the PR but it definitely feels like a footgun that relative path expression caching silently fails. I would like to make a separate PR that either:

1. Prints a warning if `nix develop`, etc. is called with a relative path expression.
2. Tries to resolves the path from CWD in the `getFingerprint` function.

Intuitively #2 sounds like less of a regression but I would appreciate guidance on which of the two options (or perhaps a different, third one) is the best.

Reply:
Some time spent learning about Nix's current eval caching behavior and looking at lazy trees later, I understand why my original suggestion doesn't make sense. Thanks for your patience. I also deleted these changes from the PR and it's ready for another look :)

> I believe a much more advanced evaluation cache could use exclusively the things that have been read as keys, but that would be more like a research project. This would make it not just capable of caching path: inputs, but also for it to be effective at caching dev shells, which tends to suffer from small changes in unrelated parts of the source.

IMO a good caching system for local flakes is very important. Relatively small flakes can take 10+ seconds to evaluate on modern hardware and this is a big barrier to relying on flakes for dev shells. I would love to learn more about what existing solutions to this are. In the meantime I can use a Bash wrapper that naively caches `nix print-dev-env` outputs.